tools/pipeline/internal/pkg/github/copy_pull_request.go

```diff
@@ -0,0 +1,468 @@
+	prBranchRef := "remotes/" + r.FromOrigin + "/" + prBranch
+
+	// Add our from upstream as a remote and fetch our PR branch
+	slog.Default().DebugContext(ctx, "adding CE upstream and fetching PR branch")
+	remoteRes, err := git.Remote(ctx, &libgit.RemoteOpts{
+		Command: libgit.RemoteCommandAdd,
+		Track:   []string{prBranch},
+		Fetch:   true,
+		Name:    r.FromOrigin,
+		URL:     fmt.Sprintf("https://github.com/%s/%s.git", r.FromOwner, r.FromRepo),
+	})
+	if err != nil {
+		err = fmt.Errorf("fetching target branch base ref: %s, %w", remoteRes.String(), err)
+		return res, err
+	}
+
+	// Create a new branch for our copied changes.
+	branchName := r.copyBranchNameForRef(baseRef, prBranch)
+	// We don't have local references so create a new branch from our tracking branch
+	baseBranch := "remotes/" + r.ToOrigin + "/" + baseRef
+	slog.Default().DebugContext(ctx, "checking out new copy branch")
+	checkoutRes, err := git.Checkout(ctx, &libgit.CheckoutOpts{
+		NewBranchForceCheckout: branchName, // -B
+		Branch:                 baseBranch,
+	})
+	if err != nil {
+		return res, fmt.Errorf("checking out new copy branch: %s: %w", checkoutRes.String(), err)
+	}
+
+	// Generate a merge commit message. While git is able to generate a nice merge
+	// commit with a summary of all commmit headers we create our own includes
+	// 'Co-Authored-By:' trailers in the commit message. Since we always squash
+	// all commits into a single merge commit this helps to retain attribution.
+	commits, err := listPullRequestCommits(ctx, github, r.FromOwner, r.FromRepo, int(r.PullNumber))
+	if err != nil {
+		return res, err
+	}
+
+	commitMessageFile, err := renderEmbeddedTemplateToTmpFile("copy-pr-commit-message.tmpl", struct {
+		CoAuthoredByTrailers []string
+		OriginPullRequest    *libgithub.PullRequest
+		TargetRef            string
+	}{
+		r.getCoAuthoredByTrailers(commits),
+		res.OriginPullRequest,
+		baseRef,
+	})
+	if err != nil {
+		return res, fmt.Errorf("creating merge commit message: %w", err)
+	}
+	defer func() {
+		commitMessageFile.Close()
+		_ = os.Remove(commitMessageFile.Name())
+	}()
+
+	slog.Default().DebugContext(ctx, "merging CE PR branch into new copy branch")
+	mergeRes, mergeErr := git.Merge(ctx, &libgit.MergeOpts{
+		File:     commitMessageFile.Name(),
+		NoVerify: true,
+		Strategy: libgit.MergeStrategyORT,
+		StrategyOptions: []libgit.MergeStrategyOption{
+			libgit.MergeStrategyOptionTheirs,
+			libgit.MergeStrategyOptionIgnoreSpaceChange,
+		},
+		IntoName: baseRef,
+		Commit:   prBranchRef,
+	})
+	if mergeErr != nil {
+		mergeErr = fmt.Errorf("merging CE PR branch into new copy branch: %s: %w", mergeRes.String(), mergeErr)
+	}
+
+	// If our merge failed we still want to create a pull request for our
+	// failed copy so that a manual fix can be performed.
+	if mergeErr != nil {
+		err := resetAndCreateNOOPCommit(ctx, git, baseBranch)
+		if err != nil {
+			err = errors.Join(mergeErr, err)
+		}
+
+		return res, err
```

Comment:
We don't want to continue here. We do want to reset and join any errors but continue to push and open the no-op PR.